### PR TITLE
DenseMatrix Improvements

### PR DESCRIPTION
**Pull request by ShikharJ**
Relevant: https://github.com/symengine/symengine.py/pull/143

### CODE REVIEW

**Comment by ShikharJ** [comment]
@isuruf Can you review this please?

File: symengine/matrix.h [hunk 1]
```diff
@@ -172,12 +172,12 @@ class DenseMatrix : public MatrixBase
...
-    friend void col_join(const DenseMatrix &A, const DenseMatrix &B,
-                         DenseMatrix &C);
-    friend void row_del(DenseMatrix &A, unsigned k);
-    friend void col_del(DenseMatrix &A, unsigned k);
+    virtual void row_join(const DenseMatrix &B);
+    virtual void col_join(const DenseMatrix &B);
+    virtual void row_insert(const DenseMatrix &B, unsigned pos);
+    virtual void col_insert(const DenseMatrix &B, unsigned pos);
+    virtual void row_del(unsigned k);
+    virtual void col_del(unsigned k);
```
Comment (isuruf) [inline symengine/matrix.h]: Is `virtual` needed here?

Comment (ShikharJ) [inline symengine/matrix.h]: Since other functions such as `add_matrix` is also `virtual`, I thought these should also follow suit.

Comment (isuruf) [inline symengine/matrix.h]: They are declared in `MatrixBase` and overridden in child classes. This is not.

Comment (ShikharJ) [inline symengine/matrix.h]: Updated it.

**Comment by ShikharJ** [comment]
Ping @isuruf.

File: symengine/matrix.h (hunk 1 above)
Comment (isuruf) [inline symengine/matrix.h]: I got this message. Please give the reviewers some time to review

File: symengine/dense_matrix.cpp [hunk 2]
```diff
@@ -364,76 +364,96 @@ void mul_dense_scalar(const DenseMatrix &A, const RCP<const Basic> &k,
 }
 
 // ---------------------------- Joining Operations ---------------------------//
-void row_join(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
+void DenseMatrix::row_join(const DenseMatrix &B)
 {
-    SYMENGINE_ASSERT(A.row_ == B.row_
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: Why is this removed?

File: symengine/dense_matrix.cpp [hunk 3]
```diff
@@ -364,76 +364,96 @@ void mul_dense_scalar(const DenseMatrix &A, const RCP<const Basic> &k,
...
-                     and (A.row_ == C.row_ and C.col_ == A.col_ + B.col_));
+    this->col_insert(B, col_);
+}
 
-    unsigned row = A.row_, col = A.col_;
+void DenseMatrix::col_join(const DenseMatrix &B)
+{
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: This needs a check that the cols are same

Comment (ShikharJ) [inline symengine/dense_matrix.cpp]: Since it is already checked in `row_insert` and `col_insert`, I think it will be redundant. Same for above.

File: symengine/dense_matrix.cpp [hunk 4]
```diff
@@ -364,76 +364,96 @@ void mul_dense_scalar(const DenseMatrix &A, const RCP<const Basic> &k,
...
+    this->row_insert(B, row_);
+}
 
-    for (unsigned i = 0; i < row; i++) {
+void DenseMatrix::row_insert(const DenseMatrix &B, unsigned pos)
+{
+    SYMENGINE_ASSERT(col_ == B.col_ and pos <= row_)
+
+    unsigned row = row_, col = col_;
+    DenseMatrix C = *this;
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: This can be done without a temporary. First, resize `this`, then move the elements of `this` below row `pos` to the end and then insert `B`.

File: symengine/dense_matrix.cpp [hunk 5]
```diff
@@ -364,76 +364,95 @@ void mul_dense_scalar(const DenseMatrix &A, const RCP<const Basic> &k,
...
-void col_join(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
+void DenseMatrix::col_insert(const DenseMatrix &B, unsigned pos)
 {
-    SYMENGINE_ASSERT(A.col_ == B.col_
-                     and (A.col_ == C.col_ and C.row_ == A.row_ + B.row_));
+    SYMENGINE_ASSERT(row_ == B.row_ and pos <= col_)
 
-    unsigned row = A.row_, col = A.col_;
+    unsigned row = row_, col = col_;
+    DenseMatrix C = *this;
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: Same as in `row_insert`

**Comment by ShikharJ** [comment]
Anything else left to do here @isuruf?

File: symengine/dense_matrix.cpp [hunk 6]
```diff
@@ -364,76 +364,93 @@ void mul_dense_scalar(const DenseMatrix &A, const RCP<const Basic> &k,
...
+    SYMENGINE_ASSERT(row_ == B.row_ and pos <= col_)
 
-    unsigned row = A.row_, col = A.col_;
+    unsigned row = row_, col = col_;
+    this->resize(row_, col_ + B.col_);
 
-    for (unsigned i = 0; i < row; i++) {
-        for (unsigned j = 0; j < col; j++) {
-            C.m_[i * col + j] = A.m_[i * col + j];
+    for (int i = row - 1; i >= 0; i--) {
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: maximum of `unsigned` and `int` are different.

Comment (ShikharJ) [inline symengine/dense_matrix.cpp]: But then if I change it to `unsigned`, the testing condition will always remain true.

Comment (isuruf) [inline symengine/dense_matrix.cpp]: https://stackoverflow.com/questions/5458204/unsigned-int-reverse-iteration-with-for-loops

**Comment by ShikharJ** [comment]
Thanks @isuruf.
